### PR TITLE
allow export of sha256sum for snapshots

### PR DESCRIPTION
**Pull request by LesnyRumcajs**
**Summary of changes**
Changes introduced in this pull request:
- adds an option to export a sha256 checksum after exporting the snapshot. This step adds around 20 seconds to the total export time for calibnet. Unfortunately, it would be tricky to do it with chunk hashing during snapshot creation.

**Reference issue to close (if applicable)**

**Other information and links**

Part of https://github.com/ChainSafe/forest/issues/1899

### CODE REVIEW

**Comment by lemmih** [comment]
It's tricky to compute the hash concurrently with snapshot creation before it would have to happen in the daemon rather than in the client?

**Comment by LesnyRumcajs** [comment]
> It's tricky to compute the hash concurrently with snapshot creation before it would have to happen in the daemon rather than in the client?

We would need to implement our own `AsyncWrite` as the logic is nested in https://github.com/filecoin-project/ref-fvm/blob/8d3a6148cbb65b4ec469f5115277aa6c2ab6df31/ipld/car/src/lib.rs#L31-L51 
I am not sure it's worth it though I'm happy to give it a go if you feel otherwise.

**Comment by lemmih** [comment]
> > It's tricky to compute the hash concurrently with snapshot creation before it would have to happen in the daemon rather than in the client?
> 
> We would need to implement our own `AsyncWrite` as the non-trivial logic is nested in https://github.com/filecoin-project/ref-fvm/blob/8d3a6148cbb65b4ec469f5115277aa6c2ab6df31/ipld/car/src/lib.rs#L31-L51 I am not sure it's worth it though I'm happy to give it a go if you feel otherwise.

Nah, let's leave it for the future. Or maybe @hanabi1224 could have a look at it.

**Comment by LesnyRumcajs** [comment]
> > > It's tricky to compute the hash concurrently with snapshot creation before it would have to happen in the daemon rather than in the client?
> > 
> > 
> > We would need to implement our own `AsyncWrite` as the non-trivial logic is nested in https://github.com/filecoin-project/ref-fvm/blob/8d3a6148cbb65b4ec469f5115277aa6c2ab6df31/ipld/car/src/lib.rs#L31-L51 I am not sure it's worth it though I'm happy to give it a go if you feel otherwise.
> 
> Nah, let's leave it for the future. Or maybe @hanabi1224 could have a look at it.

It shouldn't be that difficult, just a wrapper around a a `BufWriter`.

**Review by lemmih — APPROVED** [review]

**Comment by LesnyRumcajs** [comment]
@lemmih I revamped the hashing technology, please take a look again as it has changed a bit.

**Comment by LesnyRumcajs** [comment]
Something is not right after merging with mainstream, will investigate why.

**Comment by lemmih** [comment]
Can `forest_utils` and `forest_utils_sink` be merged?

We'll end up using `tokio::io::AsyncWrite` rather than `futures::AsyncWrite` but not a big deal since we can freely convert between the two.

**Comment by LesnyRumcajs** [comment]
@lemmih 

> Can `forest_utils` and `forest_utils_sink` be merged?

I can rename `forest_utils_sink` to `forest_utils_io` and move the contents of `forest_utils` there. Right now the code seems to sit in the `node/` which is a bit counterintuitive. What do you think?

**Comment by lemmih** [comment]
> @lemmih
> 
> > Can `forest_utils` and `forest_utils_sink` be merged?
> 
> I can rename `forest_utils_sink` to `forest_utils_io` and move the contents of `forest_utils` there. Right now the code seems to sit in the `node/` which is a bit counterintuitive. What do you think?

I feel like we have too may `forest_utils` crates and having a single utility crate would be much easier to handle. Right now we have `forest_net_utils`, `forest_hash_utils`, `forest_json_utils`, `forest_test_utils`, and `forest_utils`. We also have other utility crates with different names: `forest_macros` and `forest_auth`. How about we merge these into a single crate? For now, you could put your code in `forest_utils::io`.

**Comment by LesnyRumcajs** [comment]
> > @lemmih
> > > Can `forest_utils` and `forest_utils_sink` be merged?
> > 
> > 
> > I can rename `forest_utils_sink` to `forest_utils_io` and move the contents of `forest_utils` there. Right now the code seems to sit in the `node/` which is a bit counterintuitive. What do you think?
> 
> I feel like we have too may `forest_utils` crates and having a single utility crate would be much easier to handle. Right now we have `forest_net_utils`, `forest_hash_utils`, `forest_json_utils`, `forest_test_utils`, and `forest_utils`. We also have other utility crates with different names: `forest_macros` and `forest_auth`. How about we merge these into a single crate? 

I kind of prefer many smaller crates partitioned by their categories than one big crate. Makes it easier to reason about dependencies and cuts compilations times.

> For now, you could put your code in `forest_utils::io`.
This also doesn't work for me, conceptually. Importing `utils` from `node` (which `forest_utils` is actually) is a bit off for `blockchain/chain`, no? I'd recommend the approach mentioned in previous comment but if you feel strongly about this then of course I can go ahead with what you proposed.

**Comment by LesnyRumcajs** [comment]
> We'll end up using `tokio::io::AsyncWrite` rather than `futures::AsyncWrite` but not a big deal since we can freely convert between the two.

I moved it to `tokio::io::AsyncWrite`.

**Comment by lemmih** [comment]
> I kind of prefer many smaller crates partitioned by their categories than one big crate. Makes it easier to reason about dependencies and cuts compilations times.

What does it mean to "reason about dependencies" and I don't think it would cut compilation time. In what scenario would it cut compilation time?

**Comment by LesnyRumcajs** [comment]
> What does it mean to "reason about dependencies" and I don't think it would cut compilation time. 

If I would see `tide` dependency in `utils_hash` I would be really surprised and would investigate it further. Same dependency in `utils` and I wouldn't even blink because I'd just ignore the entire wall of dependencies.

> In what scenario would it cut compilation time?

To my understanding, if you change something in one significant crate dependency (especially the `Cargo.lock`), all dependents would have to rebuild. Same goes for having different binaries, you would not need to pull everything from crates.io if you'd have only small subset of dependencies.

**Comment by lemmih** [comment]
> > What does it mean to "reason about dependencies" and I don't think it would cut compilation time.
> 
> If I would see `tide` dependency in `utils_hash` I would be really surprised and would investigate it further. Same dependency in `utils` and I wouldn't even blink because I'd just ignore the entire wall of dependencies.

While nice to have, I don't think that justifies the significant administrative cost of having a plethora of crates.

> > In what scenario would it cut compilation time?
> 
> To my understanding, if you change something in one significant crate dependency (especially the `Cargo.lock`), all dependents would have to rebuild. Same goes for having different binaries, you would not need to pull everything from crates.io if you'd have only small subset of dependencies.

But the bulk of our code will use all of the utility crates. I have a hard time imagining that having two crates for `forest_hash_utils` and `forest_json_utils` could ever save any compilation time. How would we avoid recompilation by keeping those two crates separate? Do we have significant code bases that only use one crate but not the other? And do we modify those utility crates frequently enough to justify the administrative cost?

**Comment by LesnyRumcajs** [comment]
How could it not save compilation time? You change something in `metrics`, a dependency of only `forest`. Only this dependency tree will need to get checked and re-compiled.
```
❯ cargo build
   Compiling forest_metrics v0.2.0 (/home/rumcajs-work/prj/forest/utils/metrics)
   Compiling forest v0.4.0 (/home/rumcajs-work/prj/forest/forest)
    Finished dev [unoptimized] target(s) in 18.61s
```

You change something in `forest_json` which has a larger dependency tree
```
❯ cargo build
   Compiling forest_json v0.2.0 (/home/rumcajs-work/prj/forest/utils/json)
   Compiling forest_vm v0.4.0 (/home/rumcajs-work/prj/forest/vm)
   Compiling forest_networks v0.2.0 (/home/rumcajs-work/prj/forest/types/networks)
   Compiling forest_statediff v0.2.0 (/home/rumcajs-work/prj/forest/utils/statediff)
   Compiling forest_fil_types v0.4.0 (/home/rumcajs-work/prj/forest/types)
   Compiling forest_message v0.8.0 (/home/rumcajs-work/prj/forest/vm/message)
   Compiling forest_state_migration v0.2.0 (/home/rumcajs-work/prj/forest/vm/state_migration)
   Compiling forest_blocks v0.2.0 (/home/rumcajs-work/prj/forest/blockchain/blocks)
   Compiling forest_actor_interface v0.2.0 (/home/rumcajs-work/prj/forest/vm/actor_interface)
   Compiling forest_paramfetch v0.2.0 (/home/rumcajs-work/prj/forest/utils/paramfetch)
   Compiling forest_interpreter v0.2.0 (/home/rumcajs-work/prj/forest/vm/interpreter)
   Compiling forest_chain v0.2.0 (/home/rumcajs-work/prj/forest/blockchain/chain)
   Compiling forest_state_manager v0.1.0 (/home/rumcajs-work/prj/forest/blockchain/state_manager)
   Compiling forest_libp2p v0.2.0 (/home/rumcajs-work/prj/forest/node/forest_libp2p)
   Compiling forest_genesis v0.2.0 (/home/rumcajs-work/prj/forest/utils/genesis)
   Compiling forest_message_pool v0.2.0 (/home/rumcajs-work/prj/forest/blockchain/message_pool)
   Compiling forest_chain_sync v0.2.0 (/home/rumcajs-work/prj/forest/blockchain/chain_sync)
   Compiling forest_rpc-api v0.2.0 (/home/rumcajs-work/prj/forest/node/rpc-api)
   Compiling forest_fil_cns v0.2.0 (/home/rumcajs-work/prj/forest/blockchain/consensus/fil_cns)
   Compiling forest_deleg_cns v0.2.0 (/home/rumcajs-work/prj/forest/blockchain/consensus/deleg_cns)
   Compiling forest_rpc v0.2.0 (/home/rumcajs-work/prj/forest/node/rpc)
   Compiling forest_rpc-client v0.2.0 (/home/rumcajs-work/prj/forest/node/rpc-client)
   Compiling forest v0.4.0 (/home/rumcajs-work/prj/forest/forest)
    Finished dev [unoptimized] target(s) in 22.58s
```

Of course, it won't cut the time by 80%, but when you do changes and monitor everything with `cargo watch` it does matter.

**Comment by lemmih** [comment]
Your example highlights my point. Even in such an extreme case where lots of crates are recompiled, the difference is only a few seconds: 18.61s vs 22.58s.

It was good discussing the pros and cons. Let's not add any more crates at this time.

**Comment by hanabi1224** [comment]
> a few seconds: 18.61s vs 22.58s

In this case I think the linkage of forest executable takes ~15s (on my machine) for both cases, maybe `cargo check` is a better command to make the comparison.  BTW, I tried RUSTFLAGS="-Cprefer-dynamic" locally to reduce linkage time for a dev build but it failed with tons of errors. : (

**Review by elmattic — APPROVED** [review]
Good!

File: node/rpc/src/chain_api.rs [hunk 1]
```diff
@@ -70,32 +77,56 @@ where
     }
 
     let file = File::create(&out).await.map_err(JsonRpcError::from)?;
-    let writer = BufWriter::new(file);
+    let writer = AsyncWriterWithChecksum::<Sha256, _>::new(BufWriter::new(file));
```
[the remark this answers is not in the document]
Comment (lemmih) [reply node/rpc/src/chain_api.rs:80]: Hm, indeed. We'll address that in another PR.

**Pull request merged**